scp/lib/bfs/bucket/scp.rb

```diff
@@ -145,32 +145,53 @@ def mkdir_p(path)
       end
 
       def sh!(*cmd) # rubocop:disable Metrics/MethodLength
-        stdout = ''
         stderr = nil
         status = 0
         cmdstr = cmd.map {|x| Shellwords.escape(x) }.join(' ')
 
+        # init command's STDOUT stream (large output may break lines in on_data):
+        stdout_reader, stdout_writer = IO.pipe
+
         @client.session.open_channel do |ch|
           ch.exec(cmdstr) do |_, _success|
             ch.on_data do |_, data|
-              if block_given?
-                yield data
-              else
-                stdout += data
-              end
+              stdout_writer.write(data)
             end
             ch.on_extended_data do |_, _, data|
               stderr = data
             end
             ch.on_request('exit-status') do |_, buf|
               status = buf.read_long
             end
+            ch.on_close do |_|
+              stdout_writer.close
+            end
           end
         end
-        @client.session.loop
+
+        # loop in background (blocking to be done by consumer):
+        loop_thread = Thread.new { @client.session.loop }
```

Comment:
Had to move that in thread 'cause `.loop` is blocking, need to fit 2 blocking processes (loop + caller read) together.